### PR TITLE
opensc-explorer, command "info" - added information about LCS

### PR DESCRIPTION
**Pull request by popovec**
Information about "Life cycle status byte" is now available in listing.
Also src/libopensc/types.h update - added more LCSB definitions.

### CODE REVIEW

**Comment by Jakuje** [comment]
Is this actually going to do anything (except for printing the three original values of LCS)? The printing is fine, but as far as I see in the code, they are set throughout the code only to the original three values defined before and they are used as a bitmap in some cases (for some reason) so defining 0x03 overlaps with 0x01 and 0x02.

**Comment by popovec** [comment]
My aim was to add LCS values to match what is defined by the ISO7816-4 standard.. and add the functionality to   opensc-explorer to display these values under names as their ISO7816-4 defines. You are right, currently more than 3 values are not used, but in the future some previously unused value may appear on any card.

Let's look at where this value is handled bitwise:
src/libopensc/card-miocos.c
`````
        if (file->status & SC_FILE_STATUS_INVALIDATED)
                *p++ = 0;
`````
src/libopensc/card-flex.c
`````
p[11] = (file->status & SC_FILE_STATUS_INVALIDATED) ? 0x00 : 0x01;
`````
I think that we can replace it like this:
`````
  (file->status == SC_FILE_STATUS_INVALIDATED)
`````

I think it would be appropriate to rename  `SC_FILE_STATUS_INVALIDATED` to `SC_FILE_STATUS_DEACTIVATED`  as stated in ISO7816-4 .. 

And if it doesn't matter anywhere else, we  can adjust the current values of SC_FILE_STATUS_ACTIVATED,  SC_FILE_STATUS_CREATION  and SC_FILE_STATUS_INVALIDATED to match the values from ISO 7816-4

**Comment by Jakuje** [comment]
> My aim was to add LCS values to match what is defined by the ISO7816-4 standard.. and add the functionality to opensc-explorer to display these values under names as their ISO7816-4 defines.

Got that. But adding unused defines and code that can never be executed (and therefore tested) is never a good idea especially if it conflicts with usage of these defines elsewhere (bitmask discussed further).

> You are right, currently more than 3 values are not used, but in the future some previously unused value may appear on any card.

There would have to be some code that would set the status to these values (and there would have to be some value in setting them). The value is not directly read from the files but set in `iso7816_process_fci()` and couple of drivers.

> Let's look at where this value is handled bitwise:
> src/libopensc/card-miocos.c
> 
> ```
>         if (file->status & SC_FILE_STATUS_INVALIDATED)
>                 *p++ = 0;
> ```
> 
> src/libopensc/card-flex.c
> 
> ```
> p[11] = (file->status & SC_FILE_STATUS_INVALIDATED) ? 0x00 : 0x01;
> ```
> 
> I think that we can replace it like this:
> 
> ```
>   (file->status == SC_FILE_STATUS_INVALIDATED)
> ```

I think you are right. It looks like this was misunderstanding of the card driver writers.

> I think it would be appropriate to rename `SC_FILE_STATUS_INVALIDATED` to `SC_FILE_STATUS_DEACTIVATED` as stated in ISO7816-4 ..

Sounds good.

> And if it doesn't matter anywhere else, we can adjust the current values of SC_FILE_STATUS_ACTIVATED, SC_FILE_STATUS_CREATION and SC_FILE_STATUS_INVALIDATED to match the values from ISO 7816-4

I would like to avoid this. At this moment, the `SC_FILE_STATUS_` is simple three-value "enum" which can be simply compared by code in drivers. Using it as bitmask is nonsense as already said above.

If you modify it to match the values from ISO, you will get 16 values with duplicate and unused values. The rest of the library outside of `iso7816_process_fci()` should not care about this implementation detail what are values on a card and just use these three values (maybe add also the 2 more defined in the ISO and add their handling in `iso7816_process_fci()` or maybe elsewhere) so you can get this information in the opensc-explorer.

**Comment by popovec** [comment]
Yes, it is clear that `iso7816_process_fci ()` or the corresponding replacement in the card driver takes the FCI and sets the `file->status` to the appropriate symbolic name.  

I would suggest this:
`````
#define SC_FILE_STATUS_ACTIVATED	5
#define SC_FILE_STATUS_INVALIDATED	4  (of course rename here.. to DEACTIVATED)
#define SC_FILE_STATUS_CREATION         1
#define SC_FILE_STATUS_NO_INFO          0
`````
There is one more thing, for example MyEID at LCSB = 1 does not require a verify operation when accessing files / operations. ACO5 at LCSB = 3 also does not require verify. According to the comment in `src/libopensc/types.h`, SetCOS 4.4 also does not need to use verify in  SC_FILE_STATUS_CREATION state (LCSB = 1)

Also function `src/pkcs15init/pkcs15-lib.c` can be modified so that it does not require a PIN if it is not needed (based on LCSB) .. This can help us fix #1737 .. but it is not advisable to use the patch listed in https://github.com/OpenSC/OpenSC/issues/1737#issuecomment-576621305, we need one more variable for example `file->acl_inactive` (of course default value 0)  - managed by card driver -  according to the LCSB status.

**Comment by popovec** [comment]
@jakuje
You are right, it is not good to introduce functions, variables and constants that are not used.

I have updated the `iso7816_process_fci ()` ...  functionality can be tested by _default_ driver.  All other card drivers that use the  iso7816_process_fci () function  and use TAG 0x8A in accordance with the standard no longer need to use their own code. (This should be solved by another PR.)

----------------------------
It is probably necessary to clean more things about LCSB ...

If we do not use ISO7816 terminology, we can easily get to the problems..

I am not sure if here "INVALIDATED" corresponds to "Termination state"
or "Operational state (deactivated)" From ISO7816-4 this is more probably 
"Termination state" ..

card-starcos.c:
`````
                case 1:
                        status = "creation";
                        file->status = SC_FILE_STATUS_CREATION;
                        break;
                case 5:
                        status = "operational active";
                        file->status = SC_FILE_STATUS_ACTIVATED;
                        break;
                case 12:
                case 13:
                        status = "creation";
                        file->status = SC_FILE_STATUS_INVALIDATED;
                        break;
`````

If we had this globally in OpenSC, they would not have to define these 
things internally per driver, each in a different way .. and at the 
same time probably everyone adheres to ISO7816-4 ...

card-setcos.c:
`````
/* Setcos 4.4 Life Cycle Status Integer  */
#define SETEC_LCSI_CREATE      0x01
#define SETEC_LCSI_INIT        0x03
#define SETEC_LCSI_ACTIVATED   0x07
#define SETEC_LCSI_DEACTIVATE  0x06
#define SETEC_LCSI_TEMINATE    0x0F /* MF only  */


/* The ACs are always for the SETEC_LCSI_ACTIVATED state, even if
 * we have to create the file in the SC_FILE_STATUS_INITIALISATION state. */
`````


card-myeid.c
`````
#define MYEID_STATE_CREATION            0x01
#define MYEID_STATE_ACTIVATED           0x07

.  
.
                if(tag[0] == MYEID_STATE_CREATION) {
                        file->status = SC_FILE_STATUS_CREATION;
                        sc_log(card->ctx, "File id (%X) status SC_FILE_STATUS_CREATION (0x%X)",
                                        file->id, tag[0]);
                }
                else if(tag[0] == MYEID_STATE_ACTIVATED) {
                        file->status = SC_FILE_STATUS_ACTIVATED;
                        sc_log(card->ctx, "File id (%X) status SC_FILE_STATUS_ACTIVATED (0x%X)",
                                        file->id, tag[0]);
                }

`````

card-epass2003.c
`````
                if (tag[0] == 0x01)
                        file->status = SC_FILE_STATUS_CREATION;
                else if (tag[0] == 0x07 || tag[0] == 0x05)
                        file->status = SC_FILE_STATUS_ACTIVATED;
                else if (tag[0] == 0x06 || tag[0] == 0x04)
                        file->status = SC_FILE_STATUS_INVALIDATED;
`````

**Review by Jakuje — APPROVED** [review]
Thanks for updating the PR. Now it looks fine. I added just one note.

If this could be more widely used by drivers without the need to reimplement it again, it would be great. Unfortunately, I don't have much iso cards.

File: src/libopensc/types.h [hunk 1]
```diff
@@ -225,10 +225,27 @@ typedef struct sc_acl_entry {
...
+#define SC_FILE_STATUS_PROPRIETARY	0xf0 /* ISO7816-4: codes > 15 */
+
+/* reserved for future use by ISO/IEC */
+#define SC_FILE_STATUS_RFU_2		0x07 /* ISO7816-4: (0x02) */
+#define SC_FILE_STATUS_RFU_8		0x08 /* ISO7816-4: (0x08) */
+#define SC_FILE_STATUS_RFU_9		0x09 /* ISO7816-4: (0x09) */
+#define SC_FILE_STATUS_RFU_10		0x0a /* ISO7816-4: (0x0a) */
+#define SC_FILE_STATUS_RFU_11		0x0b /* ISO7816-4: (0x0b) */
+
+#define SC_FILE_STATUS_UNKNOWN		0xff /* if tag 0x8A is missing, there is no information about LCSB */
```
[the remark this answers is not in the document]
Comment (popovec) [reply src/libopensc/types.h:247]: SC_FILE_STATUS_NO_INFO  - The tag 0x8A is present in FCI, but value 0x00 indicates, that card does not provide information about LCS. Same for FCP - if I create a file, I set the 0x8A tag  to SC_FILE_STATUS_NO_INFO,  card is responsible to change this to real life cycle status or of course can accept 0x00 as is.

SC_FILE_STATUS_UNKNOWN - if tag 0x8A is missing in FCI ..

**Comment by frankmorgner** [comment]
thanks for clarification and review

**Pull request merged**